### ISSUE TITLE
Data-driven TRX files are incompatible with Microsoft Test Manager (MTM)

### ISSUE THREAD

**Issue by TroyWalshProf** [Microsoft/vstest#894]
## Description
> When data-driven test results are upload to MTM (via the VSTS test agent or TCM.exe) you only ever get the first data-driven result.  
>
>The problem is that the current tooling for uploading test results was **built against MSTest**. MSTest and VSTest both produce TRX files, but the **formats are different**, especially when it comes to how they structure data driven results. MSTest TRX files have “InnerResults” which are children of “UnitTestResult” and hold the individual data driven result for each iteration.VSTest does not do this; VSTest just makes a “UnitTestResult” for each data driven iteration. 

Dumbed down MSTest TRX example:
```
    <UnitTestResult testName="DRIVENTEST" resultType="DataDrivenTest">
     <Counters error="0" failed="0" timeout="0" aborted="0" inconclusive="0" passedButRunAborted="0" notRunnable="0" notExecuted="0" disconnected="0" warning="0" passed="2" completed="0" inProgress="0" pending="0" />
     <InnerResults>
        <UnitTestResult testName="DRIVENTEST (Data Row 0)" outcome="Passsed" dataRowInfo="0" resultType="DataDrivenDataRow" />
        <UnitTestResult testName="DRIVENTEST (Data Row 1)" outcome="Passsed" dataRowInfo="1" resultType="DataDrivenDataRow" />
     </InnerResults>
    </UnitTestResult>
```
Dumbed down VSTest TRX example:
```
    <UnitTestResult testName="DRIVENTEST (Data Row 0)" outcome="Passsed" />
    <UnitTestResult testName="DRIVENTEST (Data Row 1)" outcome="Passsed" />
```
## Steps to reproduce
> 1 Create a test case in a VSTS project
> 2 Add a data-driven unit test and associate it with the test case created in the previous step - Make the first test pass and second fail
> 3 In VSTS create a test plan and test suite
> 4 Add the test case to the test suite
> 5 Follow the MS instructions for running functional tests (https://www.visualstudio.com/en-us/docs/build/steps/test/run-functional-tests)
> 6 Run the tests and note that only the first passing test is recorded
*If first test fails it actually messes up MTM history
**_or_**
> 5 Instead of doing the running functional tests step you can also directly use TCM - http://bugsareinthegaps.com/uploading-automated-test-results-to-mtm-made-easy/

## Expected behavior
> All data-driven test results would get uploaded

## Actual behavior
> Only the first test result gets uploaded

## Environment
> See this problem with all versions of VSTest and every MTM version since 2010

**Comment by billros** [comment]
Nice catch Troy!  Thanks for the warning. This is a big issue!~Silently ignoring results of data driven tests?  Ouch!

**Comment by codito** [comment]
Prioritizing for next sprint since the impact is high. We should consider fixing this in TPv1 too.

**Comment by nshr12** [comment]
Hi!
Is this issue considered to be fixed soon, or should I write some bypass in my project? 
Kinda blocked by this issue :(

**Comment by nshr12** [comment]
Is this issue considered to be fix soon or should I start looking for another solutions? It`s a huge blocker.

**Comment by PBoraMSFT** [comment]
@nshr12 - addressing this is on our backlog. A holistic fix involves multiple layers of the test offering in VSTS/TFS. 
1. TRX needs to be fixed on the test platform side
2. TFS / VSTS APIs and results schema and experience needs to be updated to account for hierarchical results such as data-driven.
While addressing '1' is straightforward, we are working through '2'. At this time, I don't have an ETA to share.

**Comment by PBoraMSFT** [comment]
Closing the issue here as the enhancement is being tracked on the team's backlog.

**Comment by TroyWalshProf** [comment]
@PBoraMSFT - is the team's  backlog anywhere I can see it?
This is a problem for a lot of my projects and I really really really want to know when (if) it gets fixed.

Side note* We may different definitions for what constitutes a bug and what constitutes a enhancement ;)

**Comment by roizentner** [comment]
@PBoraMSFT I've checked your published backlog (#686). Couldn't find this issue there.

How is this issue not getting the attention it deserves? This is a major flaw that's been around since the introduction of vstest.
Now, with data-driven tests executed via VSTS, we can't be expected to rollback to the old mstest!
